services/core/Darksky/darksky/agent.py

```diff
@@ -0,0 +1,475 @@
+                * interval_minutes
+            generation_time = format_timestamp(generation_time.replace(
+                microsecond=0, second=0, minute=gen_minutes))
+        else:
+            generation_time = format_timestamp(generation_time.replace(
+                microsecond=0, second=0, minute=0))
+        for entry in response['data']:
+            entry_time = datetime.datetime.fromtimestamp(entry['time'],
+                                                         pytz.utc)
+            if 'forecast' in service:
+                data.append([json.dumps(location), generation_time, entry_time,
+                             json.dumps(entry)])
+            else:
+                data.append([json.dumps(location), entry_time, json.dumps(
+                    entry)])
+        return data
+
+    def query_current_weather(self, location):
+        """
+        Retrieve data from the Darksky API, return formatted current data and
+        store forecast data in cache
+        :param location: location dictionary requested by the user
+        :return: Timestamp and data for current data from the Darksky API
+        """
+        darksky_response = self.get_darksky_forecast(
+            SERVICES_MAPPING['SERVICE_CURRENT_WEATHER']['service'], location)
+        current_response = darksky_response.pop('currently')
+        # Darksky required attribution
+        current_response["attribution"] = "Powered by Dark Sky"
+        current_time = datetime.datetime.fromtimestamp(
+            current_response['time'], pytz.utc)
+        if not self.performance_mode:
+            # if performance mode isn't running we'll be receiving extra data
+            # that we can store to help with conserving daily api calls
+            for service in SERVICES_MAPPING:
+                if service is not 'SERVICE_CURRENT_WEATHER' and \
+                        SERVICES_MAPPING[service]['json_name'] in \
+                        darksky_response:
+                    service_response = darksky_response.pop(
+                        SERVICES_MAPPING[service]['json_name'])
+                    service_data = self.format_multientry_response(
+                        location, service_response, SERVICES_MAPPING[
+                            service]['service'])
+                    self.store_weather_records(SERVICES_MAPPING[service][
+                                                   'service'], service_data)
+        return format_timestamp(current_time), current_response
+
+    def query_forecast_service(self, service, location):
+        """
+        Generic method for requesting forecast data from the various RPC
+        forecast methods
+        :param service: forecast service type of weather data to return
+        :param location: location dictionary requested during the RPC call
+        :return: Timestamp and data returned by the Darksky weather API response
+        """
+        service_name = ''
+        for service_code in SERVICES_MAPPING:
```

Comment:
Why validate service name here? Isn’t generated by the code - i.e RPC call. Example - rpc call get_hourly_forecast(locations) - would call the downstream method with the location + correct service name ???

Reply:
This is a search for a mapped value rather than validation.

Reply:
Is there a reason there is SERVICES_MAPPING[
'SERVICE_DAILY_FORECAST'][
'service'] == “get_daily_forecast”
instead of using "get_daily_forecast" as key?